### PR TITLE
MGMT-2573 Extract and use MCO image from release image

### PR DESCRIPTION
**Pull request by sacharya**
* Extract baremetal-install binary from release.
* Get mco image from release.
* Use a mirror in case OPENSHIFT_INSTALL_RELEASE_IMAGE_MIRROR is provided
* Pass the mco image to assisted-installer

### CODE REVIEW

File: internal/host/installcmd.go [hunk 1]
```diff
@@ -179,3 +189,36 @@ func getBootDevice(log logrus.FieldLogger, hwValidator hardware.Validator, host
...
+	}()
+	_, err = ps.Write([]byte(pullSecret))
+	if err != nil {
+		return "", err
+	}
+	// flush the buffer to ensure the file can be read
+	ps.Close()
+
+	// get mco image from release image
+	cmd := exec.Command("oc", "adm", "release", "info",
```
Comment (filanov) [inline internal/host/installcmd.go:211]: is this command relevant for on prem?

Comment (sacharya) [reply internal/host/installcmd.go:211]: This is to remove the mco image hardcoded in assisted-installer and get it from release image instead. That is for both cloud and onprem usecases. For disconnected onprem, we've agreed to get the image from an environment variable

Comment (filanov) [reply internal/host/installcmd.go:211]: but here i don't see any condition that don't run this command for on-prem or did i missed something?

Comment (sacharya) [reply internal/host/installcmd.go:211]: It's a work in progress. There will be a condition for disconnected onprem.

**Comment by filanov** [comment]
subsystem failed on:
```
[2020-11-15T23:13:35.666Z] • Failure [7.590 seconds]
[2020-11-15T23:13:35.666Z] Day2 cluster tests
[2020-11-15T23:13:35.666Z] /home/workspace/assisted-service_PR-648/subsystem/day2_cluster_test.go:23
[2020-11-15T23:13:35.666Z]   check installation - node registers after reboot [It]
[2020-11-15T23:13:35.667Z]   /home/workspace/assisted-service_PR-648/subsystem/day2_cluster_test.go:287
[2020-11-15T23:13:35.667Z] 
[2020-11-15T23:13:35.667Z]   Expected
[2020-11-15T23:13:35.667Z]       <int>: 0
[2020-11-15T23:13:35.667Z]   to be >=
[2020-11-15T23:13:35.667Z]       <int>: 1
[2020-11-15T23:13:35.667Z] 
[2020-11-15T23:13:35.667Z]   /home/workspace/assisted-service_PR-648/subsystem/utils_test.go:64
```
@yevgeny-shnaidman  can you please take a look? http://assisted-jenkins.usersys.redhat.com/blue/organizations/jenkins/assisted-service/detail/PR-648/26/pipeline

File: internal/oc/release.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
+package oc
```
Comment (filanov) [inline internal/oc/release.go:1]: how about adding tests to this model?

File: internal/oc/release.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+// GetMCOImage gets mcoImage url from the releaseImageMirror if provided.
+// Else gets it from the source releaseImage
+func (r *release) GetMCOImage(log logrus.FieldLogger, releaseImage string, releaseImageMirror string, pullSecret string) (string, error) {
+	var mcoImage string
+	var err error
+	if releaseImageMirror != "" {
+		//TODO: Get mirror registry certificate from install-config
+		mcoImage, err = getMCOImageFromRelease(log, releaseImageMirror, pullSecret, true)
+		if err != nil {
+			log.Errorf("failed to get mco image from mirror release image %s", releaseImageMirror)
```
Comment (tsorya) [inline internal/oc/release.go:68]: lets have it with log with error. It will be much easier to debug

File: internal/oc/release.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+			return "", err
+		}
+	} else {
+		mcoImage, err = getMCOImageFromRelease(log, releaseImage, pullSecret, false)
+		if err != nil {
+			log.Errorf("failed to get mco image from release image %s", releaseImage)
```
Comment (tsorya) [inline internal/oc/release.go:74]: same

File: internal/oc/release.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	}
+	return mcoImage, err
+}
+
+func getMCOImageFromRelease(log logrus.FieldLogger, releaseImage string, pullSecret string, insecure bool) (string, error) {
+	cmd := fmt.Sprintf("oc adm release info --image-for=machine-config-operator --insecure=%t %s", insecure, releaseImage)
+	args := strings.Split(cmd, " ")
+	mcoImage, err := execute(log, pullSecret, args[0], args[1:]...)
+	if err != nil {
+		log.Errorf("error running \"oc adm release info\" for release %s", releaseImage)
```
Comment (tsorya) [inline internal/oc/release.go:86]: same

File: internal/oc/release.go (hunk 3 above)
Comment (sacharya) [reply internal/oc/release.go]: Done

File: internal/oc/release.go (hunk 4 above)
Comment (sacharya) [reply internal/oc/release.go]: Done

File: internal/oc/release.go (hunk 5 above)
Comment (tsorya) [reply internal/oc/release.go:86]: log.WithError?

File: internal/oc/release.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	err := os.MkdirAll(workdir, 0755)
+	if err != nil {
+		return "", err
+	}
+
+	cmd := fmt.Sprintf("oc adm release extract --command=openshift-baremetal-install --to=%s --insecure=%t %s", workdir, insecure, releaseImage)
+	args := strings.Split(cmd, " ")
+	_, err = execute(log, pullSecret, args[0], args[1:]...)
+	if err != nil {
+		log.Errorf("error running \"oc adm release extract\" for release %s", releaseImage)
```
Comment (tsorya) [inline internal/oc/release.go:128]: log.WithError

File: internal/oc/release.go (hunk 2 above)
Comment (sacharya) [reply internal/oc/release.go:1]: I will add more tests in a followup PR.

Comment (filanov) [reply internal/oc/release.go:1]: Why not in this one? 
We don't push code without test.

Comment (sacharya) [reply internal/oc/release.go:1]: Added unit test. PTAL.

File: internal/host/instructionmanager_test.go [hunk 7]
```diff
@@ -192,6 +195,7 @@ func checkStepsByState(state string, host *models.Host, db *gorm.DB, mockEvents
 		{DriveType: "disk", Name: "sdh", SizeBytes: validDiskSize},
 	}
 	mockValidator.EXPECT().GetHostValidDisks(gomock.Any()).Return(disks, nil).AnyTimes()
+	mockRelease.EXPECT().GetMCOImage(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return("mcoImage", nil).AnyTimes()
```
Comment (filanov) [inline internal/host/instructionmanager_test.go:198]: Bad habit to use `AnyTimes` and probably need to add a case with a failure.

File: internal/oc/release_test.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+var (
+	log                = logrus.New()
+	releaseImage       = "release_image"
+	releaseImageMirror = "release_image_mirror"
+	cacheDir           = "/tmp"
+	pullSecret         = "pull secret"
+)
+
+var _ = Describe("oc", func() {
```
Comment (filanov) [inline internal/oc/release_test.go:22]: please add bad flows as well.

Comment (sacharya) [reply internal/oc/release_test.go:22]: Added.

File: internal/host/instructionmanager_test.go (hunk 7 above)
Comment (sacharya) [reply internal/host/instructionmanager_test.go:198]: This is inside a generic helper method to check state, that is called from many testcases.

**Review by filanov — APPROVED** [review]

**Comment by romfreiman** [comment]
@sacharya @mhrivnak @celebdor why this was not merged yet?

**Review by filanov — APPROVED** [review]

**Review by filanov — APPROVED** [review]

**Pull request merged**